pkg/compactor/executor.go

```diff
@@ -0,0 +1,313 @@
+					level.Info(e.logger).Log("msg", "job canceled by scheduler, stopping work", "job_id", jobId, "tenant", jobTenant)
+					cancelJob(err) // Cancel the job context to stop the main work
+					return
+				}
+				level.Warn(e.logger).Log("msg", "failed to send keep-alive update", "job_id", jobId, "tenant", jobTenant, "err", err)
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// sendFinalJobStatus sends a final status update to the scheduler with retry policy.
+// Compaction jobs send final statuses on completion, planning jobs only on failure for reassignment.
+func (e *schedulerExecutor) sendFinalJobStatus(ctx context.Context, key *schedulerpb.JobKey, spec *schedulerpb.JobSpec, status schedulerpb.UpdateType) {
+	jobId := key.Id
+	jobTenant := spec.Tenant
+
+	var err error
+	switch spec.JobType {
+	case schedulerpb.COMPACTION:
+		req := &schedulerpb.UpdateCompactionJobRequest{Key: key, Tenant: spec.Tenant, Update: status}
+		err = failsafe.NewExecutor(e.retryPol).WithContext(ctx).Run(func() error {
+			_, err := e.schedulerClient.UpdateCompactionJob(ctx, req)
+			return err
+		})
+	case schedulerpb.PLANNING:
+		req := &schedulerpb.UpdatePlanJobRequest{Key: key, Update: status}
+		err = failsafe.NewExecutor(e.retryPol).WithContext(ctx).Run(func() error {
+			_, err := e.schedulerClient.UpdatePlanJob(ctx, req)
+			return err
+		})
+	default:
+		err = fmt.Errorf("unsupported job type %q, only COMPACTION and PLANNING are supported", spec.JobType.String())
+	}
+
+	if err != nil {
+		level.Error(e.logger).Log("msg", "failed to send final status update", "job_id", jobId, "tenant", jobTenant, "status", status, "err", err)
+	}
+}
+
+func (e *schedulerExecutor) makeSchedulerClient() (schedulerpb.CompactorSchedulerClient, *grpc.ClientConn, error) {
+	invalidClusterReporter := util.NewInvalidClusterValidationReporter(e.cfg.GRPCClientConfig.ClusterValidation.Label, e.invalidClusterValidation, e.logger)
+	opts, err := e.cfg.GRPCClientConfig.DialOption(nil, nil, invalidClusterReporter)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	opts = append(opts, grpc.WithStatsHandler(otelgrpc.NewClientHandler()))
+	// nolint:staticcheck // grpc.Dial() has been deprecated; we'll address it before upgrading to gRPC 2.
+	conn, err := grpc.Dial(e.cfg.SchedulerAddress, opts...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	client := schedulerpb.NewCompactorSchedulerClient(conn)
+	return client, conn, nil
+}
+
+func (e *schedulerExecutor) leaseAndExecuteJob(ctx context.Context, c *MultitenantCompactor, workerID string) (bool, error) {
+	req := &schedulerpb.LeaseJobRequest{
+		WorkerId: workerID,
+	}
+
+	resp, err := e.schedulerClient.LeaseJob(ctx, req)
+	if err != nil {
+		return false, err
+	}
+
+	jobID := resp.Key.Id
+	jobTenant := resp.Spec.Tenant
+	jobType := resp.Spec.JobType
+
+	// Create a cancellable context for this job that can be canceled if the scheduler cancels the job
+	jobCtx, cancelJob := context.WithCancelCause(ctx)
+	defer cancelJob(nil)
```

Comment:
`defer` is too late since it could be sending in progress updates all the way until after the final status is already sent. The right place would be after `executeCompactionJob` or `executePlanningJob` returns (regardless of whether it's an error or not).